reactor-core/src/main/java/reactor/core/publisher/FluxPublish.java

```diff
@@ -578,9 +663,193 @@ public Object scanUnsafe(Attr key) {
 		}
 
+		static void clearAndFinalize(PublishSubscriber<?> instance) {
+			for (; ; ) {
+				final long state = instance.state;
+
+				if (isFinalized(state)) {
+					instance.clear();
+					return;
+				}
+
+				if (isSubscriptionSet(state)) {
+					instance.clear();
+				}
+
+				if (STATE.compareAndSet(
+						instance, state,
+						(state & ~WORK_IN_PROGRESS_MASK) | FINALIZED_FLAG)) {
+					break;
+				}
+			}
+		}
+
+		static long addWork(PublishSubscriber<?> instance) {
+			for (;;) {
+				long state = instance.state;
+
+				if (STATE.compareAndSet(instance, state, addWork(state))) {
+					return state;
+				}
+			}
+		}
+
+		static long addWorkIfSubscribed(PublishSubscriber<?> instance) {
+			for (;;) {
+				long state = instance.state;
+
+				if (!isSubscriptionSet(state)) {
+					return state;
+				}
+
+				if (STATE.compareAndSet(instance, state, addWork(state))) {
+					return state;
+				}
+			}
+		}
+
+		static long addWork(long state) {
+			if ((state & WORK_IN_PROGRESS_MASK) == WORK_IN_PROGRESS_MASK) {
+				return (state &~ WORK_IN_PROGRESS_MASK) | 1;
+			}
+			else {
+				return state + 1;
+			}
+		}
+
+		static long markTerminated(PublishSubscriber<?> instance) {
+			for (;;) {
+				long state = instance.state;
+
+				if (isCancelled(state) || isTerminated(state)) {
+					return state;
+				}
+
+				long nextState = addWork(state);
+				if (STATE.compareAndSet(instance, state, nextState | TERMINATED_FLAG)) {
+					return state;
+				}
+			}
+		}
+
+		static long markConnected(PublishSubscriber<?> instance) {
+			for (;;) {
+				long state = instance.state;
+
+				if (isConnected(state)) {
+					return state;
+				}
+
+				if (STATE.compareAndSet(instance, state, state | CONNECTED_FLAG)) {
+					return state;
+				}
+			}
+		}
+
+		static long markSubscriptionSet(PublishSubscriber<?> instance) {
+			for (;;) {
+				long state = instance.state;
+
+				if (isCancelled(state)) {
+					return state;
+				}
+
+				if (STATE.compareAndSet(instance, state, state | SUBSCRIPTION_SET_FLAG)) {
+					return state;
+				}
+			}
+		}
+
+		static long markSubscriptionSetAndAddWork(PublishSubscriber<?> instance) {
+			for (;;) {
+				long state = instance.state;
+
+				if (isCancelled(state)) {
+					return state;
+				}
+
+				long nextState = addWork(state);
+				if (STATE.compareAndSet(instance, state, nextState | SUBSCRIPTION_SET_FLAG)) {
+					return state;
+				}
+			}
+		}
+
+		static long markCancelled(PublishSubscriber<?> instance) {
+			for (;;) {
+				long state = instance.state;
+
+				if (isCancelled(state)) {
+					return state;
+				}
+
+				long nextState = addWork(state);
+				if (STATE.compareAndSet(instance, state, nextState | CANCELLED_FLAG)) {
+					return state;
+				}
+			}
+		}
+
+		static long markWorkDone(PublishSubscriber<?> instance, long expectedState) {
+			for (;;) {
+				long state = instance.state;
+
+				if (expectedState != state) {
+					return state;
+				}
+
```

Comment:
What is the caller expected to do in this case? The loop keeps repeating unless a cancelled state was recorded. Can different transitions occur in the meantime? Shouldn't the drain loop check them and update the newly added parameters?

Reply:
if we observe that expectedState is not equal to real we return new state which is stored in expectedState parameter within drain loop. if it is cancelled we clearandfinalize and exit if it is something else (e.g. new consumer, or new demand or new value) the `for (;;) ` loop is circle again, read at the very beginning new parameters and tries to go over all the statements within drain loop